js/source/geojson_source.js

```diff
@@ -34,9 +38,30 @@ GeoJSONSource.prototype = util.inherit(Source, {
         return this;
     },
 
-    update() {
-        if (this._dirty) this._updateData();
-        if (this._loaded) this._updateTiles();
+    onAdd(map) {
+        this.map = map;
+    },
+
+    loaded() {
+        return this._loaded && this._pyramid.loaded();
+    },
+
+    update(transform) {
+        if (this._dirty) {
+            this._updateData();
+        }
+
+        if (this._loaded) {
+            this._pyramid.update(this.used, transform);
+        }
+    },
+
+    render(layers, painter) {
+        if (this._loaded) {
+            this._pyramid.renderedIDs().forEach((id) => {
```

Comment:
I seem to recall that when I was playing with my little bike rides
visualization, V8 wouldn't optimize render() because it had a forEach. So
it might be a bigger performance bottleneck than it seems, at least in
Chrome.

On Friday, January 9, 2015, Vladimir Agafonkin <notifications@github.com>
wrote:

> In js/source/geojson_source.js
> <https://github.com/mapbox/mapbox-gl-js/pull/925#discussion-diff-22723879>
> :
>
> > +        return this._loaded && this._pyramid.loaded();
> > +    },
> > +
> > +    update(transform) {
> > +        if (this._dirty) {
> > +            this._updateData();
> > +        }
> > +
> > +        if (this._loaded) {
> > +            this._pyramid.update(this.used, transform);
> > +        }
> > +    },
> > +
> > +    render(layers, painter) {
> > +        if (this._loaded) {
> > +            this._pyramid.renderedIDs().forEach((id) => {
>
> maybe lets avoid forEach in sensitive functions like render? Changing this
> to a simple loop may not make a big difference but small things like this
> add up.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mapbox/mapbox-gl-js/pull/925/files#r22723879>.
>